File: node/fe/ui/pages/wallet.js
'use strict';

/**
 * Page object for the mini-cex HTML pages. The server renders small, labelled
 * pages (a class per figure), so the page object reads by label and the
 * assertions are about the exchange, not about the markup.
 *
 * The base URL is the running service; a page that never loads (service down)
 * surfaces as ScreenNotReady, which the steps turn into Blocked.
 */

const BASE = (process.env.MINI_CEX_URL || 'http://127.0.0.1:8110').replace(/\/+$/, '');

class ScreenNotReady extends Error {}

class WalletPage {
  constructor(page) { this.page = page; this.base = BASE; }

  async open(path = '/') {
    try { await this.page.goto(this.base + path, { waitUntil: 'domcontentloaded', timeout: 15_000 }); }
    catch (err) { throw new ScreenNotReady('mini-cex page ' + path + ' did not load: ' + err.message); }
  }

  /** Home: one row per asset, with symbol, USD price text, and listed/delisted label. */
  async assets() {
    await this.page.waitForSelector('ul.assets li.asset', { timeout: 15_000 }).catch(() => { throw new ScreenNotReady('asset list never rendered'); });
    return this.page.$$eval('ul.assets li.asset', (els) => els.map((el) => ({
      symbol: el.getAttribute('data-symbol'),
      symbolText: el.querySelector('.symbol') ? el.querySelector('.symbol').textContent.trim() : '',
      usdText: el.querySelector('.usd') ? el.querySelector('.usd').textContent.trim() : '',
      activeText: el.querySelector('.active') ? el.querySelector('.active').textContent.trim() : '',
    })));
  }

  /** Wallet: one row per balance, with asset and amount text (whole coins). */
  async balances() {
    await this.page.waitForSelector('h1.handle', { timeout: 15_000 }).catch(() => { throw new ScreenNotReady('wallet page never rendered'); });
    return this.page.$$eval('ul.balances li.balance', (els) => els.map((el) => ({
      asset: el.getAttribute('data-asset'),
      amountText: el.querySelector('.amount') ? el.querySelector('.amount').textContent.trim() : '',
    })));
  }

  async op() {
    await this.page.waitForSelector('h1.op-id', { timeout: 15_000 }).catch(() => { throw new ScreenNotReady('op page never rendered'); });
    const read = async (sel) => (await this.page.$(sel)) ? (await this.page.$eval(sel, (e) => e.textContent.trim())) : null;
    return { idText: await read('h1.op-id'), directionText: await read('.direction'), statusText: await read('.status'), amountText: await read('.amount') };
  }

  /** Open a write form (/forms/transfer|swap|bridge). */
  async openForm(action) {
    await this.open('/forms/' + action);
    await this.page.waitForSelector('#form', { timeout: 15_000 }).catch(() => { throw new ScreenNotReady(action + ' form never rendered'); });
  }
  /** Fill fields by id, submit, and wait for the result line the handler writes. */
  async submitForm(values) {
    for (const [id, val] of Object.entries(values)) await this.page.fill('#' + id, String(val));
    await this.page.click('#go');
    await this.page.waitForSelector('#result[data-status]', { timeout: 15_000 }).catch(() => { throw new ScreenNotReady('form result never appeared'); });
  }
  async formResult() {
    return {
      status: Number(await this.page.getAttribute('#result', 'data-status')),
      code: await this.page.getAttribute('#result', 'data-code'),
      text: (await this.page.textContent('#result')) || '',
    };
  }
}

module.exports = { WalletPage, ScreenNotReady, BASE };

File: python/fe/ui/pages/wallet.py
"""Page object for the mini-cex HTML pages. Mirror of node/fe/ui/pages/wallet.js.
Reads by label; a page that never loads raises ScreenNotReady (grades Blocked)."""

from __future__ import annotations

import os

BASE = os.environ.get("MINI_CEX_URL", "http://127.0.0.1:8110").rstrip("/")


class ScreenNotReady(Exception):
    pass


class WalletPage:
    def __init__(self, page) -> None:
        self.page = page
        self.base = BASE

    def open(self, path="/"):
        try:
            self.page.goto(self.base + path, wait_until="domcontentloaded", timeout=15000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady(f"mini-cex page {path} did not load: {err}") from err

    def assets(self):
        try:
            self.page.wait_for_selector("ul.assets li.asset", timeout=15000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady("asset list never rendered") from err
        return self.page.eval_on_selector_all("ul.assets li.asset", """els => els.map(el => ({
            symbol: el.getAttribute('data-symbol'),
            symbolText: el.querySelector('.symbol') ? el.querySelector('.symbol').textContent.trim() : '',
            usdText: el.querySelector('.usd') ? el.querySelector('.usd').textContent.trim() : '',
            activeText: el.querySelector('.active') ? el.querySelector('.active').textContent.trim() : '',
        }))""")

    def balances(self):
        try:
            self.page.wait_for_selector("h1.handle", timeout=15000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady("wallet page never rendered") from err
        return self.page.eval_on_selector_all("ul.balances li.balance", """els => els.map(el => ({
            asset: el.getAttribute('data-asset'),
            amountText: el.querySelector('.amount') ? el.querySelector('.amount').textContent.trim() : '',
        }))""")

    def op(self):
        try:
            self.page.wait_for_selector("h1.op-id", timeout=15000)
        except Exception as err:  # noqa: BLE001
            raise ScreenNotReady("op page never rendered") from err

        def read(sel):
            el = self.page.query_selector(sel)
            return el.text_content().strip() if el else None
        return {"idText": read("h1.op-id"), "directionText": read(".direction"), "statusText": read(".status"), "amountText": read(".amount")}
